File: examples/javascript/basics.js
/**
 * GitHub Copilot Basics - JavaScript Examples
 *
 * This file demonstrates how GitHub Copilot can generate JavaScript code
 * from natural language comments. Each section shows a prompt (comment)
 * and the code Copilot helped produce.
 */


// --- Example 1: Capitalize the first letter of each word in a string ---

function titleCase(str) {
  return str
    .split(" ")
    .map((word) => word.charAt(0).toUpperCase() + word.slice(1).toLowerCase())
    .join(" ");
}


// --- Example 2: Remove duplicate values from an array ---

function removeDuplicates(arr) {
  return [...new Set(arr)];
}


// --- Example 3: Deep clone an object ---

function deepClone(obj) {
  // Use native structuredClone when available; fall back to JSON-based clone for older runtimes.
  if (typeof structuredClone === "function") {
    return structuredClone(obj);
  }
  return JSON.parse(JSON.stringify(obj));
}


// --- Example 4: Debounce a function ---

function debounce(fn, delay) {
  let timer;
  return function (...args) {
    clearTimeout(timer);
    timer = setTimeout(() => fn.apply(this, args), delay);
  };
}


// --- Example 5: Check if a number is prime ---

function isPrime(n) {
  if (n < 2) return false;
  for (let i = 2; i <= Math.sqrt(n); i++) {
    if (n % i === 0) return false;
  }
  return true;
}


// --- Example 6: Group an array of objects by a property ---

function groupBy(arr, key) {
  return arr.reduce((groups, item) => {
    const group = item[key];
    // Guard against prototype pollution keys
    if (group === "__proto__" || group === "constructor" || group === "prototype") {
      return groups;
    }
    if (!groups[group]) {
      groups[group] = [];
    }
    groups[group].push(item);
    return groups;
  }, Object.create(null));
}


// Quick demo (Node.js)
if (typeof require !== "undefined" && require.main === module) {
  console.log(titleCase("hello world from copilot")); // Hello World From Copilot
  console.log(removeDuplicates([1, 2, 2, 3, 3, 3]));  // [1, 2, 3]
  console.log(isPrime(17));                            // true
  console.log(
    groupBy(
      [
        { name: "Alice", dept: "Engineering" },
        { name: "Bob", dept: "Design" },
        { name: "Carol", dept: "Engineering" },
      ],
      "dept"
    )
  );
}

module.exports = { titleCase, removeDuplicates, deepClone, debounce, isPrime, groupBy };

File: examples/python/basics.py
"""
GitHub Copilot Basics - Python Examples

This file demonstrates how GitHub Copilot can generate Python code from
natural language comments. Each section shows a prompt (comment) and the
code Copilot helped produce.
"""


# --- Example 1: Calculate the factorial of a number ---

def factorial(n):
    """Return the factorial of a non-negative integer n."""
    if n < 0:
        raise ValueError("n must be a non-negative integer")
    if n == 0:
        return 1
    return n * factorial(n - 1)


# --- Example 2: Check if a string is a palindrome ---

def is_palindrome(s):
    """Return True if the string s is a palindrome, ignoring case and spaces."""
    cleaned = s.replace(" ", "").lower()
    return cleaned == cleaned[::-1]


# --- Example 3: Find the most frequent element in a list ---

def most_frequent(items):
    """Return the element that appears most frequently in the list."""
    if not items:
        return None
    return max(set(items), key=items.count)


# --- Example 4: Convert Celsius to Fahrenheit ---

def celsius_to_fahrenheit(celsius):
    """Convert a temperature from Celsius to Fahrenheit."""
    return celsius * 9 / 5 + 32


# --- Example 5: Flatten a nested list ---

def flatten(nested):
    """Recursively flatten a nested list into a single list."""
    result = []
    for item in nested:
        if isinstance(item, list):
            result.extend(flatten(item))
        else:
            result.append(item)
    return result


if __name__ == "__main__":
    print(factorial(5))                        # 120
    print(is_palindrome("racecar"))            # True
    print(most_frequent([1, 2, 2, 3, 3, 3]))  # 3
    print(celsius_to_fahrenheit(100))          # 212.0
    print(flatten([1, [2, [3, 4]], 5]))        # [1, 2, 3, 4, 5]
